### ISSUE TITLE
Allow package from other sources other than GitHub

### ISSUE THREAD

**Issue by f-f** [purescript/pursuit#422]
Moving over here from https://github.com/purescript/registry/issues/15, which mentions that the only problems of allowing packages to be published from other sources other than GitHub come from Pursuit relying on GitHub for some things. Quoting from there:

> Right now Pursuit only supports packages hosted on GitHub, and nobody has complained [..]. The reasons for this are that allows us to use the GitHub API to get a rendered HTML readme (although I now consider that a misfeature and I'd like to remove it), and also it allows us to construct source links, since the source links take you to GitHub. If we allow non-GH packages in the registry, this means we won't be able to upload those packages to Pursuit right now (which is not a dealbreaker but probably worth considering). I'd quite like to have package sources hosted in the same place as the HTML API docs for source links though, which will probably be easier to do now with our shiny new registry; if we did that we'd be able to allow non-GH packages without losing source links.

**Comment by toastal** [comment]
For Sourcehut you can get users and README in one fell swoop through their GraphQL API ([schema](https://git.sr.ht/~sircmpwn/git.sr.ht/tree/master/item/api/graph/schema.graphqls)).

```graphql
query {
 repositoryByName(name: "purescript-web-page-visibility") {
    readme
    owner { id }
    description
  }
  user(username: "toastal") {
    id
    url
  }
}
```

Which returns something like

```json
{
  "data": {
    "repositoryByName": {
      "readme": "$ESCAPED_README_AS_HTML",
      "description": "Type definitions and low level interface implementations for W3C Page Visibility API",
      "owner": {
        "id": 11755
      }
    },
    "user": {
      "id": 11755,
      "url": "https://toast.al"
    }
  }
```

Querying through cURL is something like

```console
$ oauth_token=your oauth token
$ curl \
  --oauth2-bearer "$oauth_token" \
  -H 'Content-Type: application/json' \
  -d '{"query": "{ version { major, minor, patch } }"}' \
  https://meta.sr.ht/query
```

with [full documentation](https://man.sr.ht/graphql.md).

**Comment by hdgarrood** [comment]
Rather than maintaining numerous integrations, I’d rather have Pursuit know as little as possible about the various code hosting services there are. In particular I would rather move in the direction of having the JSON upload contain Markdown which will be rendered by Pursuit itself instead of using APIs to render readmes. See https://github.com/purescript/pursuit/issues/150#issuecomment-271272657

**Comment by toastal** [comment]
Oh, I wouldn't disagree, but now you definitely need to have your own sanitizer. One of my quibbles with GitHub's Markdown renderer, is that it removes `<abbr>` tags that add to accessibility, but if you let in script tags and who knows what else, that could be a concern (I assume this was 100% already thought about though).

---

One thing I appreciate about Sourcehut is that through their GraphQL API, you can `POST` any sort of HTML -- BYOR, bring your own renderer. Similar to not relying on a singular Git forge, you're also not relying on a markup language. Markdown has no specific spec really and is quite limiting for many application. In my own case, I used ran `asciidoctor` from my [Nix Flake package](https://git.sr.ht/~toastal/purescript-web-page-visibility/tree/v1.0.1/item/flake.nix#L70-92) and had the [build system](https://git.sr.ht/~toastal/purescript-web-page-visibility/tree/v1.0.1/item/.build.yml#L27-42) cURL `POST` the rendered HTML output. I would personally say BYOR is preferable.

**Comment by hdgarrood** [comment]
Yes, we already have a sanitiser, for rendering the markdown in doc-comments. The problem with allowing people to bring their own renderer is that you then have to account for any kind of dodgy HTML they might produce. We can already sanitise markdown, but sanitising HTML is harder.

**Comment by toastal** [comment]
Then will you bring a Markdown flavor that supports tables, abbreviations, admonitions, imports, and more out-of-the-box features? That's my biggest concern more than syntax A vs syntax B.

**Comment by hdgarrood** [comment]
No - this is just for the package’s readme on Pursuit. If you need all those things, I’d suggest setting up a separate documentation site.

**Comment by toastal** [comment]
Ah, that is reasonable :vulcan_salute:
